### DISCUSSION TITLE
Using query builder and filter by morphTo relationship not works.

### DISCUSSION THREAD

**Discussion by mobasherfasihi** [spatie/laravel-query-builder#813] [General]
I have the following model
```
class MailJob extends Model
{
    /**
     * Get the parent mailable model
     */
    public function mailable()
    {
        return $this->morphTo();
    }

   public function notCompleted()
   {
      return $this->whereHasMorph('mailable', User::class, fn ($q) => $q->whereNotNull('name'));  
   }

}
```

And then when I want to filter by mailable, it gives `Subject class "Illuminate\Database\Query\Builder" is invalid.`

```
QueryBuilder::for(MailJob::class)
     ->notCompleted()
    ->get();
```

**Comment by leandrodiogenes** [comment]
Hi @mobasherfasihi .
```php
QueryBuilder::for(MailJob::class)
```
that line returns an new instance of 'Spatie\QueryBuilder\QueryBuilder'.
notCompleted is an public method of your model, not of the querybuilder.

Please tell me what version of the package and laravel you are using.

**Comment by leandrodiogenes** [comment]
"notCompleted()" method seems to be an scope.
Laravel docs recommends to make a different approach.
[https://laravel.com/docs/9.x/eloquent#local-scopes](https://laravel.com/docs/9.x/eloquent#local-scopes)

```php
class MailJob extends Model
{
    public function mailable()
    {
        return $this->morphTo();
    }

   public function scopeNotCompleted($query)
   {
      return $query->whereHasMorph('mailable', fn ($q) => $q->whereNotNull('name'));  
   }

}

...

MailJob::notCompleted()->get()

```

now you can try this..
```php
QueryBuilder::for(MailJob::notCompleted())->get();
```

**Reply by mobasherfasihi** [reply]
@leandrodiogenes Thanks for your suggestion, yeah that method is scope method, just I use builder class and transferred that scope method from model to builder.
I changed my model back to the way you suggested, but still face same issue.

Just that whereHasMorph should be like this:
   
```

    public function scopeNotCompleted($query)
    {
        return $query->whereHasMorph('mailable', User::class, fn ($q) => $q->whereNotNull('name'));
    }

```

I removed event the queryBuilder, and used like `MailJob::whereHasMorph('mailable', User::class, fn($q) => $q->whereNotNull('name'))`, still same issue, it looks like whereHasMorph is working properly in my case.

But strange thing is, that I've commented the spatie query builder, but this error still point out that its from spatie query builder class.

**Reply by mobasherfasihi** [reply]
Fixed, my mistake was extending from spatie builder instead of eloquent.

Thanks for help @leandrodiogenes
